---
title: "Multi-omics analysis of histological types of breast cancer"
output: html_notebook
---

# Dataset preparation

## Loading of clinical data

Loading of clinical data from excel table
```{r}
clinicaldata.breast = read.csv("data/clinical/clinical", sep = "\t")
clinicaldata.breast
```


```{r}
clinicaldata.breast.filtered = clinicaldata.breast[ , c("sampleID", "Age_at_Initial_Pathologic_Diagnosis_nature2012",
                                                                 "histological_type", "pathologic_stage")]
na.omit(clinicaldata.breast.filtered)
```


```{r}
library(ggplot2)
clinicaldata.breast.filtered |> 
  ggplot(aes(`histological_type`, fill = `histological_type`)) + 
  geom_bar() + 
  ggtitle('Classes of histological_type') + 
  geom_text(aes(label = after_stat(count)), stat = 'count', vjust = 0) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
```

```{r}
clinicaldata.breast.filtered |> 
  ggplot(aes(`pathologic_stage`, fill = `pathologic_stage`)) + 
  geom_bar() + 
  ggtitle('Classes of pathologic_stage') + 
  geom_text(aes(label = after_stat(count)), stat = 'count', vjust = 0) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
```

```{r}
clinicaldata.breast.filtered |> 
  ggplot(aes(`Age_at_Initial_Pathologic_Diagnosis_nature2012`)) + 
  geom_bar() + 
  ggtitle('Classes of Age_at_Initial_Pathologic_Diagnosis_nature2012')
```

Filter samples only for two histology types:
```{r}
histology_names = c("Infiltrating Ductal Carcinoma", "Infiltrating Lobular Carcinoma")
clinicaldata.breast.filtered =  clinicaldata.breast.filtered[clinicaldata.breast.filtered$histological_type %in% histology_names, , drop = FALSE]
clinicaldata.breast.filtered$histological_type = chartr(" ", "_", clinicaldata.breast.filtered$histological_type)
clinicaldata.breast.filtered
```

Combine stages and delete groups with small number of samples
```{r}
pathologic_stage = c("Stage I", "Stage II", "Stage III")
clinicaldata.breast.filtered$pathologic_stage[clinicaldata.breast.filtered$pathologic_stage == "Stage IA" | clinicaldata.breast.filtered$pathologic_stage == "Stage IB"] = "Stage I"
clinicaldata.breast.filtered$pathologic_stage[clinicaldata.breast.filtered$pathologic_stage == "Stage IIA" | clinicaldata.breast.filtered$pathologic_stage == "Stage IIB"] = "Stage II"
clinicaldata.breast.filtered$pathologic_stage[clinicaldata.breast.filtered$pathologic_stage == "Stage IIIA" | 
                                         clinicaldata.breast.filtered$pathologic_stage == "Stage IIIB" | 
                                         clinicaldata.breast.filtered$pathologic_stage == "Stage IIIC"] = "Stage III"
clinicaldata.breast.filtered =  clinicaldata.breast.filtered[clinicaldata.breast.filtered$pathologic_stage %in% pathologic_stage, ]
na.omit(clinicaldata.breast.filtered)
clinicaldata.breast.filtered = clinicaldata.breast.filtered[order(clinicaldata.breast.filtered$histological_type),]
na.omit(clinicaldata.breast.filtered)
clinicaldata.breast.filtered
```

```{r}
library(ggplot2)
clinicaldata.breast.filtered |> 
  ggplot(aes(`histological_type`, fill = `histological_type`)) + 
  geom_bar() + 
  ggtitle('Classes of histological_type') + 
  geom_text(aes(label = after_stat(count)), stat = 'count', vjust = 0) 
```

```{r}
sampleIDs = clinicaldata.breast.filtered$sampleID
head(sampleIDs)
```

How many samples we have from clinical data for BRCA?
```{r}
length(sampleIDs)
```



## Feature counts filtering


Reading the second table from raw data set
```{r}
feature_count_24 = read.delim("data/RAW RNA-seq/GSM1536837_06_01_15_TCGA_24.tumor_Rsubread_FeatureCounts.txt", header = TRUE, sep = "\t", dec = ".")
```

Transform sample IDs from clinical data to merge it with RNA read counts data
```{r}
sampleIDs.trans =  gsub("-", ".", sampleIDs)
head(sampleIDs.trans)
```

```{r}
names.use <- names(feature_count_24)[(substr(names(feature_count_24), 1,15)%in% sampleIDs.trans)]
feature_count_24.subset <- feature_count_24[, names.use]
feature_count_24.subset
```


Transform data frame before saving in file
```{r}
feature_count_24.subset = cbind(row.names = feature_count_24$X, feature_count_24.subset)
new_colunm_names = colnames(feature_count_24.subset)
new_colunm_names = chartr(".", "-", new_colunm_names)
colnames(feature_count_24.subset) = new_colunm_names
feature_count_24.subset
```

```{r}
feature_count_24.ids = chartr("-", ".",colnames(feature_count_24.subset))
head(feature_count_24.ids)
```


## Methylation data filtering

Load nethylation data
```{r}
breast = read.delim("data/RAW Methylation data/breast/methy", header = TRUE, sep = " ", dec = ".")
breast
```

Check if the samples id from breast data exist in sample ids list that we need
```{r}
names.use <- names(breast)[(names(breast) %in% substr(feature_count_24.ids, 1, 15))]
breast.subset <- breast[, names.use]
breast.subset
```

How many samples we have from methylation data?
```{r}
ncol(breast.subset)
```


## Merge multi-omics data (RNA expression, methylation and clinical data) and save in file

```{r}
names.use <- names(feature_count_24.subset)[(substr(chartr("-", ".",(names(feature_count_24.subset))), 1,15)%in% names(breast.subset))]
feature_count_24.subset.breast <- feature_count_24.subset[, names.use]
feature_count_24.subset.breast
```

```{r}
names.use <- clinicaldata.breast.filtered$sampleID[((chartr("-", ".", clinicaldata.breast.filtered$sampleID)) %in% names(breast.subset))]
clinicaldata.breast.filtered <- clinicaldata.breast.filtered[clinicaldata.breast.filtered$sampleID %in% names.use, ]
clinicaldata.breast.filtered
```


Export transcriptomics data to file
```{r}
write.table(feature_count_24.subset, file = "data/filtered/RNA_FeatureCounts_filtered.txt", sep = "\t",
            row.names = TRUE, col.names = TRUE, quote=FALSE)
```

Export epigenetics data to file
```{r}
write.table(breast.subset, file = "data/filtered/Methylation_data_BRCA_filtered.txt", sep = "\t",
            row.names = TRUE, col.names = TRUE, quote=FALSE)
```

Export clinical data to file
```{r}
write.table(clinicaldata.breast.filtered, file = "data/filtered/Clinical_breast_filtered.txt", sep = "\t",
            row.names = TRUE, col.names = TRUE, quote=FALSE)
```

```{r}
library(ggplot2)
clinicaldata.breast.filtered |> 
  ggplot(aes(`histological_type`, fill = `histological_type`)) + 
  geom_bar() + 
  ggtitle('Classes of histological_type') + 
  geom_text(aes(label = after_stat(count)), stat = 'count', vjust = 0) 
```

# Transcriptomics data analysis

## Differential expression analysis

Differential expression analysis allows us to test tens of thousands of hypotheses (one test for each gene) against the null hypothesis that the activity of the gene stays the same in two different conditions. There are multiple limiting factors that influence the power of detecting genes that have real changes between two biological conditions. Among these are the limited number of biological replicates, non-normality of the distribution of the read counts, and higher uncertainty of measurements for lowly expressed genes than highly expressed genes.

There, we calculated gene-wise variability and shrunk each gene-wise variability towards the median variability of all genes. In the case of RNA-seq the dispersion coefficient $\alpha$ is shrunk towards the value of dispersion from other genes with similar read counts.

Load raw read counts
```{r}
countData = read.table("data/filtered/RNA_FeatureCounts_filtered.txt", header=TRUE, sep="\t", as.is=c(1,2))
countData
```

Load clinical data
```{r}
colData = read.table("data/filtered/Clinical_breast_filtered.txt", header=TRUE, sep="\t", as.is=c(1,2))
samples = colData$sampleID
colData = colData[-1]
rownames(colData) = samples
colData
```

```{r}
# extract column names of the data frames
cols_df_1 <- paste(chartr("-", ".", rownames(colData)),'A', sep ="")
cols_df_2 <- substr((names(countData)), 1 , 16)

# find the intersection of both column name vectors
cols_difference <- setdiff(cols_df_1, cols_df_2)
cols_intersection <- intersect(cols_df_1, cols_df_2)

cols_difference2 = paste(substr((cols_difference), 1 , 15), "B", sep = "")
cols_intersection2 <- intersect(cols_difference2, cols_df_2)
cols_intersection2
```


```{r}
cols_intersection = c(cols_intersection, cols_difference2)

```


```{r}
countData.filtered = countData[substr((names(countData)), 1 , 16) %in% cols_intersection]

# subset the initial data frames
drop_columns = c("TCGA.A7.A13D.01A.13R.A277.07", "TCGA.A7.A13E.01A.11R.A277.07", "TCGA.A7.A26E.01A.11R.A277.07", "TCGA.A7.A26J.01A.11R.A277.07")
countData.filtered = countData.filtered[ -which(names(countData.filtered) %in% drop_columns)]

#countData = countData.filtered
names(countData.filtered) = chartr(".", "-", substr(names(countData.filtered), 1 , 15))
countData = countData.filtered
```

Adjust samples to count data
```{r}
countData = countData[ , names(countData) %in% rownames(colData)]
countData = countData[,match(rownames(colData), colnames(countData))]

```

```{r}
countData
```
Save countData in file

```{r}
write.table(countData, file = "data/filtered/RNA_RAW_FeatureCounts_filtered.txt", sep = "\t",
            row.names = TRUE, col.names = TRUE, quote=FALSE)
```

```{r}
library(ggplot2)
colData |> 
  ggplot(aes(`histological_type`, fill = `histological_type`)) + 
  geom_bar() + 
  ggtitle('Classes of histological_type') + 
  geom_text(aes(label = after_stat(count)), stat = 'count', vjust = 0)
```

```{r}
colData |> 
  ggplot(aes(`pathologic_stage`, fill = `pathologic_stage`)) + 
  geom_bar() + 
  ggtitle('Classes of pathologic_stage') + 
  geom_text(aes(label = after_stat(count)), stat = 'count', vjust = 0)
```
Undersampling
```{r}
library(dplyr)
set.seed(1)
colData.ductal = colData[colData$histological_type == "Infiltrating_Ductal_Carcinoma", ]
colData.lobular = colData[colData$histological_type == "Infiltrating_Lobular_Carcinoma", ]
colData.ductal = sample_n(colData.ductal, 177)
colData = rbind(colData.lobular, colData.ductal)
colData
```

```{r}
library(ggplot2)
colData |> 
  ggplot(aes(`histological_type`, fill = `histological_type`)) + 
  geom_bar() + 
  ggtitle('Classes of histological_type') + 
  geom_text(aes(label = after_stat(count)), stat = 'count', vjust = 0)
```
```{r}
write.table(colData, file = "data/clinical_data_354s.txt", sep = "\t",
            row.names = TRUE, col.names = TRUE, quote=FALSE)
```



```{r}
# define the design formula
designFormula <- "~ histological_type"

# Run DESeq2
library(DESeq2)
library(stats)
#create a DESeq dataset object from the count matrix and the colData 
dds <- DESeqDataSetFromMatrix(countData = countData, 
                              colData = colData, 
                              design = as.formula(designFormula))
#print dds object to see the contents
print(dds)
```
### Pre-filtering
While it is not necessary to pre-filter low count genes before running the DESeq2 functions, there are two reasons which make pre-filtering useful: by removing rows in which there are very few reads, we reduce the memory size of the dds data object, and we increase the speed of the transformation and testing functions within DESeq2. It can also improve visualizations, as features with no information for differential expression are not plotted.
```{r}
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]
```

```{r}
#For each gene, we count the total number of reads for that gene in all samples 
#and remove those that don't have at least 1 read. 
dds <- dds[ rowSums(DESeq2::counts(dds)) > 1, ]
dds <- DESeq(dds)
#compute the contrast for the 'group' variable where 'Sensitive' 
#samples are used as the control group. 
DEresults = results(dds, contrast = c("histological_type", 'Infiltrating_Lobular_Carcinoma', 'Infiltrating_Ductal_Carcinoma'))

#sort results by increasing p-value
DEresults <- DEresults[order(DEresults$pvalue),]
#shows a summary of the results
print(DEresults)
```

### MA-plot

```{r}
library(DESeq2)
DESeq2::plotMA(object = dds, ylim = c(-6, 6))
```

```{r}
resApeT <- lfcShrink(dds, coef=2, type="apeglm", lfcThreshold=1)
plotMA(resApeT, ylim=c(-11,11), cex=.8)
abline(h=c(-1,1), col="dodgerblue", lwd=2)
```

### P-value distribution

```{r}
library(ggplot2)
ggplot(data = as.data.frame(DEresults), aes(x = pvalue)) + 
  geom_histogram(bins = 100)
```
How to interpret p-value distributions: http://varianceexplained.org/statistics/interpreting-pvalue-histogram/

### PCA plot

```{r}
library(DESeq2)
vsd <- vst(dds)
#rld <- rlog(dds)
```

```{r}
pcaData = DESeq2::plotPCA(vsd, ntop = 600, intgroup = c('histological_type'), returnData = TRUE) 
DESeq2::plotPCA(vsd, ntop = 600, intgroup = 'histological_type') +  
  ylim(-50, 50) + theme_bw() + geom_point(size=0.5) + xlim(-50, 50)
```


### Volcano plot

```{r}
library(ggplot2)
library(ggrepel)
data = as.data.frame(DEresults)
#Add a column to the data frame to specify if they are UP- or DOWN- regulated (log2fc respectively positive or negative)
data$diffexpressed <- "NO"
# if log2Foldchange > 0.6 and pvalue < 0.05, set as "UP"
data$diffexpressed[data$log2FoldChange > 0.6 & data$pvalue < 0.05] = "UP"
# if log2Foldchange < -0.6 and pvalue < 0.05, set as "DOWN"
data$diffexpressed[data$log2FoldChange < -0.6 & data$pvalue < 0.05] = "DOWN"
data$gene_symbol = rownames(data)
# Create a new column "delabel" to de, that will contain the name of the top 30 differentially expressed genes (NA in case they are not)
data$delabel <- ifelse(data$gene_symbol %in% head(data[order(data$padj), "gene_symbol"], 30), data$gene_symbol, NA)


# Explore a bit
head(data[order(data$padj) & data$diffexpressed == 'DOWN', ])

theme_set(theme_classic(base_size = 12) +
            theme(
              axis.title.y = element_text(face = "bold", margin = margin(0,20,0,0), size = rel(1.1), color = 'black'),
              axis.title.x = element_text(hjust = 0.5, face = "bold", margin = margin(20,0,0,0), size = rel(1.1), color = 'black'),
              plot.title = element_text(hjust = 0.5)
            ))
# Create a basic volcano plot
ggplot(data = data, aes(x = log2FoldChange, y = -log10(pvalue),  col = diffexpressed, label = delabel)) +
  geom_vline(xintercept = c(-0.6, 0.6), col = "gray", linetype = 'dashed') +
  geom_hline(yintercept = -log10(0.05), col = "gray", linetype = 'dashed') + 
  geom_point(size = 0.5) +
  scale_color_manual(values = c("#00AFBB", "grey", "#FFDB6D"), # to set the colours of our variable
                     # to set the labels in case we want to overwrite the categories from the dataframe
                    labels = c("Downregulated", "Not significant", "Upregulated")) +
  geom_text_repel(max.overlaps = Inf) # To show all labels

  
```

```{r}
## plot using ggplot2
ggplot(melted_top20_sigOE) +
        geom_point(aes(x = gene, y = normalized_counts, color = sampletype)) +
        scale_y_log10() +
        xlab("Genes") +
        ylab("Normalized Counts") +
        ggtitle("Top 20 Significant DE Genes") +
        theme_bw() +
	theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
	theme(plot.title=element_text(hjust=0.5))
```


```{r fig2, fig.height = 10, fig.width = 20, fig.align = "center"}
library(pheatmap)
normalizedCounts <- DESeq2::counts(dds, normalized = TRUE)
top100genes = data$gene_symbol[data$gene_symbol %in% head(data[order(data$padj), "gene_symbol"], 100)]

M <- normalizedCounts[rownames(normalizedCounts) %in% top100genes, ]
pheatmap(log2(M+1), 
         scale = 'row', 
         cellheight=6, 
         cellwidth = 2, 
         fontsize_col = 4, 
         fontsize_row = 6,
         cluster_cols=FALSE,
         annotation_col = colData[,c("histological_type"), drop = FALSE], 
         show_rownames = TRUE,
         show_colnames = FALSE)
```
```{r}
top100genes
```
```{r}
sum(DEresults$padj < 0.05, na.rm=TRUE)
```
Save DEGs in file
```{r}
top500genes = data$gene_symbol[data$gene_symbol %in% head(data[order(data$padj), "gene_symbol"], 500)]
top500genes.normalizedCounts <- data.frame(normalizedCounts[rownames(normalizedCounts) %in% top500genes, ])
top500genes.normalizedCounts
```

```{r}
write.table(top500genes.normalizedCounts, file = "data/top500DEGs.txt", sep = "\t",
            row.names = TRUE, col.names = TRUE, quote=FALSE)
```

```{r}
top500genes.ranked = head(data[order(data$padj), "gene_symbol"], 500)
```

```{r}
write.table(top500genes.ranked, file = "data/top500DEGs_ranked.txt", sep = "\t",
            row.names = FALSE, col.names = FALSE, quote=FALSE)
```

## Functional enrichment analysis

Load selected genes by machine learning
```{r}
# load the data
ml_selected_genes <- scan(file = "data/ml_selected_genes.txt", character(), quote = "")
ml_selected_genes
```

### GO term analysis

In a typical differential expression analysis, thousands of genes are found differentially expressed between two groups of samples. While prior knowledge of the functions of individual genes can give some clues about what kind of cellular processes have been affected, e.g. by a drug treatment, manually going through the whole list of thousands of genes would be very cumbersome and not be very informative in the end. Therefore a commonly used tool to address this problem is to do enrichment analyses of functional terms that appear associated to the given set of differentially expressed genes more often than expected by chance. 

Let’s select the genes that are significantly differentially expressed between the case and control samples. Let’s extract genes that have an adjusted p-value below 0.1 and that show a 2-fold change (either negative or positive) in the case compared to control. We will then feed this gene set into the gProfileR function. The top 10 detected GO terms are displayed below.

```{r}
library(DESeq2)
library(gprofiler2)
library(knitr)

# extract differential expression results
#DEresults = results(dds, contrast = c("histological_type", 'Infiltrating_Lobular_Carcinoma', 'Infiltrating_Ductal_Carcinoma'))
#remove genes with NA values 
#DE <- DEresults[!is.na(DEresults$padj),]
#select genes with adjusted p-values below 0.1
#DE <- DE[DE$padj < 0.1,]
#select genes with absolute log2 fold change above 1 (two-fold change)
#DE <- DE[abs(DE$log2FoldChange) > 1,]

#get the list of genes of interest
#enesOfInterest <- rownames(DE)

#calculate enriched GO terms
#goResults <- gprofiler(query = genesOfInterest, 
#                     organism = 'hsapiens', 
#                     src_filter = 'GO', 
#                     hier_filtering = 'moderate')

goResults <- gost(query = ml_selected_genes, 
                organism = "hsapiens", ordered_query = FALSE, 
                multi_query = FALSE, significant = TRUE, exclude_iea = FALSE, 
                measure_underrepresentation = FALSE, evcodes = TRUE, 
                user_threshold = 0.05, correction_method = "g_SCS", 
                domain_scope = "annotated", custom_bg = NULL, 
                numeric_ns = "", sources = NULL, as_short_link = FALSE, highlight = TRUE)

```


```{r}
#Let's define the first gene set as the list of genes from one of the
#significant GO terms found in the GO analysis. order go results by pvalue
goResults.results <- goResults$result[order(goResults$result$p_value),]
#restrict the terms that have at most 100 genes overlapping with the query
go <- goResults.results[goResults.results$intersection_size < 100,]
# use the top term from this table to create a gene set 
geneSet1 <- unlist(strsplit(go[1,]$intersection, ','))

#Define another gene set by just randomly selecting 25 genes from the counts
#table get normalized counts from DESeq2 results
normalizedCounts <- DESeq2::counts(dds, normalized = TRUE)
geneSet2 <- sample(rownames(normalizedCounts), 25)

geneSets <- list('top_GO_term' = geneSet1,
                 'random_set' = geneSet2)
```

```{r}
geneSets
```


```{r}
library(gage)
#use the normalized counts to carry out a GSEA. 
ref = match(rownames(colData[colData$histological_type == 'Infiltrating_Lobular_Carcinoma',]), 
                       colnames(normalizedCounts))
samp = match(rownames(colData[colData$histological_type == 'Infiltrating_Ductal_Carcinoma',]), 
                        colnames(normalizedCounts))
gseaResults <- gage(exprs = log2(normalizedCounts + 1), 
           ref =  ref,
           samp = samp,
           gsets = geneSets, compare = 'as.group')
gseaResults
```

We can see that the random gene set shows no significant up- or down-regulation, while the gene set we defined using the top GO term shows a significant down-regulation (adjusted p-value < 0.00016)

It is worthwhile to visualize these systematic changes in a heatmap 

```{r fig2, fig.height = 5, fig.width = 25, fig.align = "center"}
library(pheatmap)
library(RColorBrewer)
library(viridis)
# get the expression data for the gene set of interest
M <- normalizedCounts[rownames(normalizedCounts) %in% geneSet1, ]
#M = M[ ,match(rownames(colData), colnames(M))]
# log transform the counts for visualization scaling by row helps visualizing
# relative change of expression of a gene in multiple conditions
pheatmap(log2(M+1),
         annotation_col = colData[c("histological_type")], 
         show_rownames = TRUE, 
         fontsize_row = 14,
         scale = 'row', 
         cutree_cols =2, 
         cutree_rows = 2,
         cluster_cols = TRUE,
         fontsize_col = 3,
         cellheight=20)
```


```{r}
library(DOSE)
data(geneSet1)
#de <- names(geneSet1)[abs(geneSet1) > 2]

edo <- enrichDGN(geneSet1)
```
https://yulab-smu.top/biomedical-knowledge-mining-book/enrichplot.html

https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7859841/

https://www.mdpi.com/1422-0067/23/18/10967

# Machine learning

```{r}
library(DESeq2)

```


